libraries/AP_HAL_SITL/Scheduler.cpp

```diff
@@ -209,29 +216,98 @@ void Scheduler::stop_clock(uint64_t time_usec)
 */
 void *Scheduler::thread_create_trampoline(void *ctx)
 {
-    AP_HAL::MemberProc *t = (AP_HAL::MemberProc *)ctx;
-    (*t)();
-    free(t);
+    struct thread_attr *a = (struct thread_attr *)ctx;
+    a->f[0]();
+    
+    WITH_SEMAPHORE(_thread_sem);
+    if (threads == a) {
+        threads = a->next;
+    } else {
+        for (struct thread_attr *p=threads; p->next; p=p->next) {
+            if (p->next == a) {
+                p->next = p->next->next;
+                break;
+            }
+        }
+    }
+    free(a->stack);
+    free(a->f);
+    delete a;
     return nullptr;
 }
 
+#ifndef PTHREAD_STACK_MIN
+#define PTHREAD_STACK_MIN 16384U
+#endif
 
 /*
   create a new thread
 */
 bool Scheduler::thread_create(AP_HAL::MemberProc proc, const char *name, uint32_t stack_size, priority_base base, int8_t priority)
 {
-    // take a copy of the MemberProc, it is freed after thread exits
-    AP_HAL::MemberProc *tproc = (AP_HAL::MemberProc *)malloc(sizeof(proc));
-    if (!tproc) {
-        return false;
-    }
-    *tproc = proc;
+    WITH_SEMAPHORE(_thread_sem);
+
+    // even an empty thread takes 2500 bytes on Linux, so always add 2300, giving us 200 bytes
+    // safety margin
+    stack_size += 2300;
+    
     pthread_t thread {};
-    if (pthread_create(&thread, NULL, thread_create_trampoline, tproc) != 0) {
-        free(tproc);
+    uint32_t alloc_stack = MAX(PTHREAD_STACK_MIN,stack_size);
+    
+    struct thread_attr *a = new struct thread_attr;
+    if (!a) {
         return false;
     }
+    // take a copy of the MemberProc, it is freed after thread exits
+    a->f = (AP_HAL::MemberProc *)malloc(sizeof(proc));
+    if (!a->f) {
+        goto failed;
+    }
+    a->stack = memalign(4096, alloc_stack);
+    if (!a->stack) {
+        goto failed;
+    }
+    memset(a->stack, stackfill, alloc_stack);
+    a->stack_min = (const uint8_t *)((((uint8_t *)a->stack) + alloc_stack) - stack_size);
+
+    a->stack_size = stack_size;
+    a->f[0] = proc;
+    a->name = name;
+    
+    pthread_attr_init(&a->attr);
+    if (pthread_attr_setstack(&a->attr, a->stack, alloc_stack) != 0) {
+        AP_HAL::panic("Failed to set stack of size %u for thread %s", alloc_stack, name);
+    }
+    if (pthread_create(&thread, &a->attr, thread_create_trampoline, a) != 0) {
+        goto failed;
+    }
+    a->next = threads;
+    threads = a;
     return true;
+
+failed:
+    if (a->stack) {
+        free(a->stack);
+    }
+    if (a->f) {
+        free(a->f);
+    }
+    delete a;
+    return false;
 }
 
+/*
+  check for stack overflow
+ */
+void Scheduler::check_thread_stacks(void)
+{
+    WITH_SEMAPHORE(_thread_sem);
+    for (struct thread_attr *p=threads; p; p=p->next) {
+        const uint8_t ncheck = 8;
+        for (uint8_t i=0; i<ncheck; i++) {
+            if (p->stack_min[i] != stackfill) {
+                AP_HAL::panic("stack overflow in thread %s\n", p->name);
```

Comment:
There's general agreement amongst the HALs that the `\n` is inferred - ChibiOS is an exception.